File: bundle-deployer/src/types/bundle.ts
// Bundle data structures

export interface Bundle {
  base_url: string;
  dashboard: Dashboard;
  other_dashboards?: Dashboard[];
  alert_rules?: AlertRules;
  method: string;
  method_overrides?: MethodOverrides;
  name: string;
  source: string;
  beta: boolean;
  tables: Table[];
  summary_tables?: SummaryTable[];
  ui: Ui;
  metadata: Metadata;
  dependencies?: Dependencies;
}

export interface Dashboard {
  path: string;
  project_var: string;
  sha256?: string;
}

export interface AlertRules {
  path: string;
  sha256?: string;
}

export interface MethodOverrides {
  region?: string;
  stream_prefix?: string;
}

export interface Table {
  dashboard_var: string;
  name: string;
  transforms: Transform[];
}

export interface SummarySqlFile {
  path: string;
  sha256?: string;
}

export interface SummaryTable {
  name: string;
  parent_table_name: string;
  dashboard_var: string;
  sql: SummarySqlFile;
}

export interface Transform {
  path: string;
  sha256?: string;
  sample?: string;
}

export interface Ui {
  primary_url: string;
  method: Graphics;
  source: Graphics;
  data_category: string;
}

export interface Graphics {
  full_title: string;
  icon_url: string;
}

export interface Metadata {
  version: string;
  maintainer: string;
  description: string;
  channel_type: string;
}

export interface Dependencies {
  grafana?: GrafanaDependencies;
  hydrolix?: HydrolixDependencies;
  "data-sources"?: DataSource[];
}

export interface GrafanaDependencies {
  version?: string;
  plugins?: GrafanaPlugin[];
}

export interface GrafanaPlugin {
  name: string;
  version: string;
}

export interface HydrolixDependencies {
  cluster_version?: string;
  required_dictionaries?: Dictionary[];
  required_functions?: Function[];
}

export interface Dictionary {
  name: string;
  source: string;
}

export interface Function {
  name: string;
  definition: string;
}

export interface DataSource {
  name: string;
  type: string;
  url: string;
  access: string;
}

export interface Output {
  cluster_domain: string;
  project_name: string;
  grafana_domain: string;
  datalink: string;
  dashboard_id: string;
  tables: OutputTable[];
}

export interface OutputTable {
  table_name: string;
  transforms: OutputTransformation[];
}

export interface OutputTransformation {
  name: string;
  data_type: string;
  data_sub_type: string;
}

// Validation functions for bundle structure
export function validateBundle(bundle: Bundle): void {
  validateHttpsUrl(bundle.base_url, "base_url");
  validateHttpsUrl(bundle.ui.primary_url, "ui.primary_url");
  
  validateMethod(bundle.method);
  validateSource(bundle.source);
  validateName(bundle.name);
  validateChannelType(bundle.metadata.channel_type);
  validateDataCategory(bundle.ui.data_category);
  
  validateDashboard(bundle.dashboard);
  
  if (bundle.other_dashboards) {
    bundle.other_dashboards.forEach(dash => validateDashboard(dash));
  }
  
  if (bundle.alert_rules) {
    validateAlertRules(bundle.alert_rules);
  }
  
  bundle.tables.forEach(table => {
    validateMacroName(table.dashboard_var);
    table.transforms.forEach(transform => {
      validateUrlPath(transform.path);
      if (transform.sha256) validateSha256(transform.sha256);
      if (transform.sample) validateUrlPath(transform.sample);
    });
  });
  
  if (bundle.summary_tables) {
    bundle.summary_tables.forEach(sumTable => {
      validateMacroName(sumTable.dashboard_var);
      validateUrlPath(sumTable.sql.path);
      if (sumTable.sql.sha256) validateSha256(sumTable.sql.sha256);
    });
  }
  
  validateHttpsUrl(bundle.ui.method.icon_url, "ui.method.icon_url");
  validateHttpsUrl(bundle.ui.source.icon_url, "ui.source.icon_url");
  
  if (bundle.dependencies?.hydrolix?.required_dictionaries) {
    bundle.dependencies.hydrolix.required_dictionaries.forEach(dict => {
      // Dictionaries can be local paths OR URLs, so don't validate as strict HTTPS
      // Just check it's not empty
      if (!dict.source || dict.source.trim().length === 0) {
        throw new Error(`dictionary.${dict.name}.source cannot be empty`);
      }
    });
  }
}

function validateDashboard(dashboard: Dashboard): void {
  validateUrlPath(dashboard.path);
  if (dashboard.sha256) validateSha256(dashboard.sha256);
}

function validateAlertRules(alertRules: AlertRules): void {
  validateUrlPath(alertRules.path);
  if (alertRules.sha256) validateSha256(alertRules.sha256);
}

function validateHttpsUrl(url: string, fieldName: string): void {
  try {
    new URL(url);
  } catch {
    throw new Error(`${fieldName}: Failed to parse URL`);
  }
  
  if (!url.startsWith("https://") && !url.startsWith("file://")) {
    throw new Error(`${fieldName}: URL must start with https:// or file://`);
  }
}

function validateUrlPath(path: string): void {
  if (path.startsWith("/") || path.includes("..")) {
    throw new Error("Path cannot start with slash or contain ..");
  }
  
  const lower = path.toLowerCase();
  if (!lower.endsWith(".json") && !lower.endsWith(".tsv") && !lower.endsWith(".sql")) {
    throw new Error("Path must end in .json, .tsv, or .sql");
  }
}

function validateMacroName(name: string): void {
  if (name.length < 5) {
    throw new Error(`${name} Must be in format __VARIABLE_NAME__`);
  }
  
  if (!name.startsWith("__") || !name.endsWith("__")) {
    throw new Error(`${name} Must be in format __VARIABLE_NAME__`);
  }
  
  const inner = name.slice(2, -2);
  
  if (inner.length === 0) {
    throw new Error(`${name} Empty inside __VARIABLE_NAME__`);
  }
  
  if (!/^[A-Z0-9_]+$/.test(inner)) {
    throw new Error(`${name} Must be upper-case or a numeral __VARIABLE_NAME__`);
  }
  
  if (inner.includes("__")) {
    throw new Error(`${name} Inside there are no double underscores`);
  }
}

function validateMethod(method: string): void {
  const validMethods = ["firehose", "s3", "kinesis", "lambda", "http_streaming", "http"];
  if (!validMethods.includes(method)) {
    throw new Error(`${method} is an invalid method`);
  }
}

function validateSource(source: string): void {
  if (!/^[a-zA-Z0-9_-]+$/.test(source)) {
    throw new Error(
      `${source} contains invalid characters (only alphanumeric, dashes, and underscores allowed)`
    );
  }
}

function validateName(name: string): void {
  if (!/^[a-zA-Z0-9_-]+$/.test(name)) {
    throw new Error(
      `${name} contains invalid characters (only alphanumeric, dashes, and underscores allowed)`
    );
  }
}

function validateChannelType(channelType: string): void {
  const validTypes = ["AWS", "Azure", "GCP", "3rdParty", "Internal"];
  if (!validTypes.includes(channelType)) {
    throw new Error(`${channelType} is an invalid channel_type`);
  }
}

function validateDataCategory(category: string): void {
  const validCategories = ["video", "cdn", "security"];
  if (!validCategories.includes(category)) {
    throw new Error(`${category} is an invalid data category`);
  }
}

function validateSha256(hash: string): void {
  if (hash.length !== 64) {
    throw new Error(`${hash} needs to be 64 characters long`);
  }
  if (!/^[0-9a-fA-F]+$/.test(hash)) {
    throw new Error(`${hash} must contain only hexadecimal characters`);
  }
}

File: src/models/bundle.rs
use serde::{de, Deserialize, Serialize};
use url::Url;

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(deny_unknown_fields)]
pub struct Bundle {
    #[serde(deserialize_with = "deserialize_valid_name")]
    pub name: String,
    #[serde(deserialize_with = "deserialize_valid_source")]
    pub source: String,
    #[serde(deserialize_with = "deserialize_valid_method")]
    pub method: String,
    pub method_overrides: Option<MethodOverrides>,
    #[serde(default)]
    pub solution: bool,
    pub beta: bool,
    #[serde(deserialize_with = "deserialize_https_url")]
    pub base_url: String,
    pub dashboard: Dashboard,
    pub other_dashboards: Option<Vec<Dashboard>>,
    pub tables: Vec<Table>,
    pub summary_tables: Option<Vec<SummaryTable>>,
    pub ui: Ui,
    pub metadata: Metadata,
    #[serde(default)]
    pub dependencies: Option<Dependencies>,
    pub alert_rules: Option<AlertRules>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(deny_unknown_fields)]
pub struct Dashboard {
    #[serde(deserialize_with = "deserialize_url_path")]
    pub path: String,
    pub project_var: String,
    #[serde(default, deserialize_with = "deserialize_optional_sha256")]
    pub sha256: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(deny_unknown_fields)]
pub struct AlertRules {
    #[serde(deserialize_with = "deserialize_url_path")]
    pub path: String,
    #[serde(default, deserialize_with = "deserialize_optional_sha256")]
    pub sha256: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(deny_unknown_fields)]
pub struct MethodOverrides {
    pub region: Option<String>,
    pub stream_prefix: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(deny_unknown_fields)]
pub struct Table {
    #[serde(deserialize_with = "deserialize_macro_name")]
    pub dashboard_var: String,
    pub name: String,
    pub transforms: Vec<Transform>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(deny_unknown_fields)]
pub struct SummarySqlFile {
    #[serde(deserialize_with = "deserialize_url_path")]
    pub path: String,
    #[serde(default, deserialize_with = "deserialize_optional_sha256")]
    pub sha256: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(deny_unknown_fields)]
pub struct SummaryTable {
    pub name: String,
    pub parent_table_name: String,
    #[serde(deserialize_with = "deserialize_macro_name")]
    pub dashboard_var: String,
    pub sql: SummarySqlFile,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(deny_unknown_fields)]
pub struct Transform {
    #[serde(deserialize_with = "deserialize_url_path")]
    pub path: String,
    #[serde(default, deserialize_with = "deserialize_optional_sha256")]
    pub sha256: Option<String>,
    #[serde(default, deserialize_with = "deserialize_optional_url_path")]
    pub sample: Option<String>,
    pub method: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(deny_unknown_fields)]
pub struct Ui {
    #[serde(deserialize_with = "deserialize_https_url")]
    pub primary_url: String,
    pub method: Graphics,
    pub source: Graphics,
    #[serde(deserialize_with = "deserialize_valid_data_categories")]
    pub data_category: String,
    #[serde(default, deserialize_with = "deserialize_optional_folder")]
    pub folder: Option<String>,
    #[serde(default, deserialize_with = "deserialize_optional_subfolder")]
    pub subfolder: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(deny_unknown_fields)]
pub struct Graphics {
    pub full_title: String,
    #[serde(deserialize_with = "deserialize_https_url")]
    pub icon_url: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(deny_unknown_fields)]
pub struct Metadata {
    pub version: String,
    pub maintainer: String,
    pub description: String,
    #[serde(deserialize_with = "deserialize_valid_channel_type")]
    pub channel_type: String,
}

// New structs for dependencies
#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(deny_unknown_fields)]
pub struct Dependencies {
    pub grafana: Option<GrafanaDependencies>,
    pub hydrolix: Option<HydrolixDependencies>,
    #[serde(default)]
    #[serde(rename = "data-sources")]
    pub data_sources: Option<Vec<DataSource>>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(deny_unknown_fields)]
pub struct GrafanaDependencies {
    pub version: Option<String>,
    pub plugins: Option<Vec<GrafanaPlugin>>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(deny_unknown_fields)]
pub struct GrafanaPlugin {
    pub name: String,
    pub version: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(deny_unknown_fields)]
pub struct HydrolixDependencies {
    #[serde(rename = "cluster_version")]
    pub cluster_version: Option<String>,
    #[serde(rename = "required_dictionaries")]
    pub required_dictionaries: Option<Vec<String>>,
    #[serde(rename = "required_functions")]
    pub required_functions: Option<Vec<String>>,
    #[serde(rename = "shared_dictionaries")]
    pub shared_dictionaries: Option<Vec<String>>,
    #[serde(rename = "shared_functions")]
    pub shared_functions: Option<Vec<String>>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(deny_unknown_fields)]
pub struct DataSource {
    pub name: String,
    #[serde(rename = "type")]
    pub data_source_type: String,
    pub url: String,
    pub access: String,
}

// Custom deserializer for HTTPS URLs
fn deserialize_https_url<'de, D>(deserializer: D) -> Result<String, D::Error>
where
    D: de::Deserializer<'de>,
{
    let s = match String::deserialize(deserializer) {
        Ok(v) => v,
        Err(e) => {
            return Err(de::Error::custom(format!(
                "failed to deserialize url: {}",
                e
            )))
        }
    };

    // Validate URL format
    match Url::parse(&s) {
        Ok(_) => (),
        Err(e) => return Err(de::Error::custom(format!("Failed to parse URL: {}", e))),
    }

    // Ensure HTTPS
    if !s.starts_with("https://") && !s.starts_with("file://") {
        return Err(de::Error::custom("URL must start with https:// or file://"));
    }
    Ok(s)
}

// Custom deserializer for HTTP path
fn deserialize_url_path<'de, D>(deserializer: D) -> Result<String, D::Error>
where
    D: de::Deserializer<'de>,
{
    let s = match String::deserialize(deserializer) {
        Ok(v) => v,
        Err(e) => {
            return Err(de::Error::custom(format!(
                "failed to deserialize url path: {}",
                e
            )))
        }
    };

    if s.starts_with("/") || s.contains("..") {
        return Err(de::Error::custom(
            "Path cannot start with slash or contain 2 dots".to_string(),
        ));
    }

    if !s.to_lowercase().ends_with(".json")
        && !s.to_lowercase().ends_with(".tsv")
        && !s.to_lowercase().ends_with(".sql")
    {
        return Err(de::Error::custom(
            "Path must end in .json or .tsv or .sql".to_string(),
        ));
    }

    Ok(s)
}

// Custom __VARIABLE__ name
fn deserialize_macro_name<'de, D>(deserializer: D) -> Result<String, D::Error>
where
    D: de::Deserializer<'de>,
{
    let s = match String::deserialize(deserializer) {
        Ok(v) => v,
        Err(e) => {
            return Err(de::Error::custom(format!(
                "failed to deserialize macro name: {}",
                e
            )))
        }
    };

    // Must be at least 5 characters: "__X__"
    if s.len() < 5 {
        return Err(de::Error::custom(format!(
            "{s} Must be in format __VARIABLE_NAME__"
        )));
    }

    if !s.starts_with("__") || !s.ends_with("__") {
        return Err(de::Error::custom(format!(
            "{s} Must be in format __VARIABLE_NAME__"
        )));
    }

    // Extract the inner content (between the double underscores)
    let inner = &s[2..s.len() - 2];

    // Inner content cannot be empty
    if inner.is_empty() {
        return Err(de::Error::custom(format!(
            "{s} Empty inside __VARIABLE_NAME__"
        )));
    }

    // Check each character in the inner content
    for ch in inner.chars() {
        match ch {
            'A'..='Z' => continue, // Uppercase letters are allowed
            '1'..='9' => continue, // Uppercase letters are allowed
            '_' => continue,       // Single underscores are allowed
            _ => {
                return Err(de::Error::custom(format!(
                    "{s} Must be upper-case or a numeral __VARIABLE_NAME__"
                )));
            }
        }
    }

    // Make sure there are no consecutive underscores (double underscores)
    if inner.contains("__") {
        return Err(de::Error::custom(format!(
            "{s} Inside there are no double underscores"
        )));
    }

    Ok(s)
}

#[allow(dead_code)]
// Custom deserializer
fn deserialize_ends_in_json<'de, D>(deserializer: D) -> Result<String, D::Error>
where
    D: de::Deserializer<'de>,
{
    let s = match String::deserialize(deserializer) {
        Ok(v) => v,
        Err(e) => {
            return Err(de::Error::custom(format!(
                "failed to deserialize json path: {}",
                e
            )))
        }
    };

    match s.ends_with(".json") {
        true => Ok(s),
        false => Err(de::Error::custom(format!("{s} Must end in .json"))),
    }
}

// Custom deserializer
fn deserialize_valid_method<'de, D>(deserializer: D) -> Result<String, D::Error>
where
    D: de::Deserializer<'de>,
{
    let s = match String::deserialize(deserializer) {
        Ok(v) => v,
        Err(e) => {
            return Err(de::Error::custom(format!(
                "failed to deserialize method: {}",
                e
            )))
        }
    };

    match s.as_str() {
        // Convert String to &str for matching
        "firehose" | "s3" | "kinesis" | "lambda" | "http_streaming" | "http" | "multi_stream" => {
            Ok(s)
        }
        _ => Err(de::Error::custom(format!("{} is an invalid method", s))),
    }
}

// Custom deserializer
fn deserialize_valid_source<'de, D>(deserializer: D) -> Result<String, D::Error>
where
    D: de::Deserializer<'de>,
{
    let s = match String::deserialize(deserializer) {
        Ok(v) => v,
        Err(e) => {
            return Err(de::Error::custom(format!(
                "failed to deserialize source: {}",
                e
            )))
        }
    };

    if s.chars()
        .any(|c| !c.is_alphanumeric() && c != '-' && c != '_')
    {
        return Err(de::Error::custom(format!(
            "{} contains invalid characters (only alphanumeric, dashes, and underscores allowed)",
            s
        )));
    }

    Ok(s)
}

// Custom deserializer for channel type
fn deserialize_valid_channel_type<'de, D>(deserializer: D) -> Result<String, D::Error>
where
    D: de::Deserializer<'de>,
{
    let s = match String::deserialize(deserializer) {
        Ok(v) => v,
        Err(e) => {
            return Err(de::Error::custom(format!(
                "failed to deserialize channel type: {}",
                e
            )))
        }
    };

    match s.as_str() {
        "AWS" | "Azure" | "GCP" | "3rdParty" | "Internal" => Ok(s),
        _ => Err(de::Error::custom(format!(
            "{} is an invalid channel_type",
            s
        ))),
    }
}

// Custom deserializer
fn deserialize_valid_data_categories<'de, D>(deserializer: D) -> Result<String, D::Error>
where
    D: de::Deserializer<'de>,
{
    let s = match String::deserialize(deserializer) {
        Ok(v) => v,
        Err(e) => {
            return Err(de::Error::custom(format!(
                "failed to deserialize data category: {}",
                e
            )))
        }
    };

    match s.as_str() {
        // Convert String to &str for matching
        "video" | "cdn" | "security" => Ok(s),
        _ => Err(de::Error::custom(format!(
            "{} is an invalid data category",
            s
        ))),
    }
}

// Custom deserializer for optional SHA256
fn deserialize_optional_sha256<'de, D>(deserializer: D) -> Result<Option<String>, D::Error>
where
    D: de::Deserializer<'de>,
{
    let opt = match Option::<String>::deserialize(deserializer) {
        Ok(Some(v)) => Some(v),
        Ok(None) => None,
        Err(e) => return Err(e),
    };

    match opt {
        Some(s) => {
            if s.len() != 64 {
                Err(de::Error::custom(format!(
                    "{s} needs to be 64 characters long"
                )))
            } else if !s.chars().all(|c| c.is_ascii_hexdigit()) {
                Err(de::Error::custom(format!(
                    "{s} must contain only hexadecimal characters"
                )))
            } else {
                Ok(Some(s))
            }
        }
        None => Ok(None),
    }
}

fn deserialize_valid_name<'de, D>(deserializer: D) -> Result<String, D::Error>
where
    D: de::Deserializer<'de>,
{
    let s = match String::deserialize(deserializer) {
        Ok(v) => v,
        Err(e) => {
            return Err(de::Error::custom(format!(
                "failed to deserialize name: {}",
                e
            )))
        }
    };

    match s.as_str() {
        s_str
            if s_str
                .chars()
                .any(|c| !c.is_alphanumeric() && c != '_' && c != '-') =>
        {
            return Err(de::Error::custom(format!(
                "{} contains invalid characters (only alphanumeric characters, underscores, and dashes allowed)",
                s
            )));
        }
        _ => (),
    }

    Ok(s)
}

// Custom deserializer for optional SHA256
fn deserialize_optional_url_path<'de, D>(deserializer: D) -> Result<Option<String>, D::Error>
where
    D: de::Deserializer<'de>,
{
    let opt = match Option::<String>::deserialize(deserializer) {
        Ok(Some(v)) => Some(v),
        Ok(None) => None,
        Err(e) => return Err(e),
    };

    match opt {
        Some(s) => {
            if s.starts_with("/") || s.contains("..") {
                return Err(de::Error::custom(
                    "Path cannot start with slash or contain 2 dots".to_string(),
                ));
            }

            if !s.to_lowercase().ends_with(".json") && !s.to_lowercase().ends_with(".tsv") {
                return Err(de::Error::custom(
                    "Path must end in .json or .tsv".to_string(),
                ));
            }
            Ok(Some(s))
        }
        None => Ok(None),
    }
}

fn deserialize_optional_folder<'de, D>(deserializer: D) -> Result<Option<String>, D::Error>
where
    D: de::Deserializer<'de>,
{
    let opt = Option::<String>::deserialize(deserializer)?;
    if let Some(ref s) = opt {
        match s.as_str() {
            "api-context" | "cdn" | "dns" | "media" | "security" => {}
            _ => return Err(de::Error::custom(format!("{} is an invalid folder", s))),
        }
    }
    Ok(opt)
}

fn deserialize_optional_subfolder<'de, D>(deserializer: D) -> Result<Option<String>, D::Error>
where
    D: de::Deserializer<'de>,
{
    let opt = Option::<String>::deserialize(deserializer)?;
    if let Some(ref s) = opt {
        match s.as_str() {
            "multi-cdn" | "bots" | "ds2" | "siem" => {}
            _ => return Err(de::Error::custom(format!("{} is an invalid subfolder", s))),
        }
    }
    Ok(opt)
}

// Helper functions to extract functions and dictionaries
impl Bundle {
    /// Get all functions categorized by type (bundle-specific vs shared)
    pub fn get_all_functions(&self) -> (Vec<String>, Vec<String>) {
        let bundle_specific = self
            .dependencies
            .as_ref()
            .and_then(|d| d.hydrolix.as_ref())
            .and_then(|h| h.required_functions.as_ref())
            .cloned()
            .unwrap_or_default();

        let shared = self
            .dependencies
            .as_ref()
            .and_then(|d| d.hydrolix.as_ref())
            .and_then(|h| h.shared_functions.as_ref())
            .cloned()
            .unwrap_or_default();

        (bundle_specific, shared)
    }

    /// Get all dictionaries categorized by type (bundle-specific vs shared)
    pub fn get_all_dictionaries(&self) -> (Vec<String>, Vec<String>) {
        let bundle_specific = self
            .dependencies
            .as_ref()
            .and_then(|d| d.hydrolix.as_ref())
            .and_then(|h| h.required_dictionaries.as_ref())
            .cloned()
            .unwrap_or_default();

        let shared = self
            .dependencies
            .as_ref()
            .and_then(|d| d.hydrolix.as_ref())
            .and_then(|h| h.shared_dictionaries.as_ref())
            .cloned()
            .unwrap_or_default();

        (bundle_specific, shared)
    }
}
